### PR TITLE
NC | NSFS | Add Concurrency Tests

### PR DESCRIPTION
**Pull request by shirady**
### Explain the changes
1. Add 2 cases to our concurrency tests:
- Key with 5 versions, concurrent delete of 3 latest versions (no veID) -> 3 delete markers
- Start with 1 version, 10 Puts + 10 Reads (latest) concurrent nothing fails -> 11 version IDs
2. Refactor the file `test_versioning_concurrency` for reuse: move `tmp_fs_path` and `nsfs` as global variables, add the helper function `_upload_versions` from existing code to avoid code duplication.

### Issues: Fixed #xxx / Gap #xxx
1. In test `concurrent put object and head object latest version` notice that I increased the retries for 10 iterations (I could use the default with 5 iterations).

### Testing Instructions:
1. `sudo npx jest test_versioning_concurrency.test.js`

- [ ] Doc added/updated
- [X] Tests added

### CODE REVIEW

File: src/test/unit_tests/jest_tests/test_versioning_concurrency.test.js [hunk 1]
```diff
@@ -81,4 +77,68 @@ describe('test versioning concurrency', () => {
...
+        for (let i = 0; i < 3; i++) {
+            nsfs.delete_object({ bucket: bucket, key: key }, DUMMY_OBJECT_SDK)
+                .then(res => successful_operations.push(res))
+                .catch(err => console.log('delete latest version error - ', err));
+        }
+
+        await P.delay(1000);
+        expect(successful_operations.length).toBe(3);
+        const versions = await nsfs.list_object_versions({ bucket: bucket }, DUMMY_OBJECT_SDK);
+        expect(versions.objects.length).toBe(8); // 5 versions before + 3 delete markers concurrent
```
Comment (nadavMiz) [inline src/test/unit_tests/jest_tests/test_versioning_concurrency.test.js:98]: can this number change if previous tests put more objects?

File: src/test/unit_tests/jest_tests/test_versioning_concurrency.test.js [hunk 2]
```diff
@@ -81,4 +77,68 @@ describe('test versioning concurrency', () => {
...
+
+    it('concurrent delete of latest version', async () => {
+        const bucket = 'bucket1';
+        const key = 'key3';
+        const number_of_versions = 5;
+        const versions_arr = await _upload_versions(bucket, key, number_of_versions);
+        expect(versions_arr.length).toBe(number_of_versions);
+
+        const successful_operations = [];
+        for (let i = 0; i < 3; i++) {
```
Comment (nadavMiz) [inline src/test/unit_tests/jest_tests/test_versioning_concurrency.test.js:89]: maybe worth increasing iteration to increase risk for concurrency issue

File: src/test/unit_tests/jest_tests/test_versioning_concurrency.test.js [hunk 3]
```diff
@@ -81,4 +77,68 @@ describe('test versioning concurrency', () => {
...
+        const versions = await nsfs.list_object_versions({ bucket: bucket }, DUMMY_OBJECT_SDK);
+        expect(versions.objects.length).toBe(8); // 5 versions before + 3 delete markers concurrent
+        const delete_marker_arr = versions.objects.filter(object => object.delete_marker === true);
+        expect(delete_marker_arr.length).toBe(3);
+    });
+
+    it('concurrent put object and head object latest version', async () => {
+        const bucket = 'bucket1';
+        const key = 'key4';
+        await _upload_versions(bucket, key, 1);
```
Comment (nadavMiz) [inline src/test/unit_tests/jest_tests/test_versioning_concurrency.test.js:106]: note that you wait for nsfs.upload_object to finish. is this intended? isn't the test about put object and head object at the same time?

File: src/test/unit_tests/jest_tests/test_versioning_concurrency.test.js (hunk 2 above)
Comment (romayalon) [reply src/test/unit_tests/jest_tests/test_versioning_concurrency.test.js:89]: @shirady as long as you see a retry every time I think it's ok

File: src/test/unit_tests/jest_tests/test_versioning_concurrency.test.js [hunk 4]
```diff
@@ -81,4 +77,68 @@ describe('test versioning concurrency', () => {
...
+        const key = 'key3';
+        const number_of_versions = 5;
+        const versions_arr = await _upload_versions(bucket, key, number_of_versions);
+        expect(versions_arr.length).toBe(number_of_versions);
+
+        const successful_operations = [];
+        for (let i = 0; i < 3; i++) {
+            nsfs.delete_object({ bucket: bucket, key: key }, DUMMY_OBJECT_SDK)
+                .then(res => successful_operations.push(res))
+                .catch(err => console.log('delete latest version error - ', err));
```
Comment (romayalon) [inline src/test/unit_tests/jest_tests/test_versioning_concurrency.test.js:92]: @shirady I suggest to convert all prints to errors_operations.push(err) and after that assert that the length is 0

**Review by romayalon — APPROVED** [review]

**Pull request merged**